osquery/tables/system/windows/certificates.cpp

```diff
@@ -307,7 +337,236 @@ void parseSystemStoreString(LPCWSTR sysStoreW,
+
+  certBuff.clear();
+  getCertCtxProp(certContext, CERT_KEY_IDENTIFIER_PROP_ID, certBuff);
+  std::string subjectKeyId;
+  toHexStr(certBuff.begin(), certBuff.end(), subjectKeyId);
+  r["subject_key_id"] = subjectKeyId;
+
+  r["path"] =
+      storeLocation + "\\" + constructDisplayStoreName(storeId, storeName);
+  r["store_location"] = storeLocation;
+  r["store"] = storeName;
+
+  std::string serial;
+  toHexStr(certContext->pCertInfo->SerialNumber.pbData,
+           certContext->pCertInfo->SerialNumber.pbData +
+               certContext->pCertInfo->SerialNumber.cbData,
+           serial);
+  r["serial"] = serial;
+
+  std::string authKeyId;
+  if (certContext->pCertInfo->cExtension != 0) {
+    auto extension = CertFindExtension(szOID_AUTHORITY_KEY_IDENTIFIER2,
+                                       certContext->pCertInfo->cExtension,
+                                       certContext->pCertInfo->rgExtension);
+    if (extension != nullptr) {
+      unsigned long decodedBuffSize = 0;
+      CryptDecodeObjectEx(CERT_ENCODING,
+                          X509_AUTHORITY_KEY_ID2,
+                          extension->Value.pbData,
+                          extension->Value.cbData,
+                          CRYPT_DECODE_NOCOPY_FLAG,
+                          nullptr,
+                          nullptr,
+                          &decodedBuffSize);
+
+      certBuff.resize(decodedBuffSize, 0);
+      std::fill(certBuff.begin(), certBuff.end(), 0);
+      auto decodeRet = CryptDecodeObjectEx(CERT_ENCODING,
+                                           X509_AUTHORITY_KEY_ID2,
+                                           extension->Value.pbData,
+                                           extension->Value.cbData,
+                                           CRYPT_DECODE_NOCOPY_FLAG,
+                                           nullptr,
+                                           certBuff.data(),
+                                           &decodedBuffSize);
+      if (decodeRet != FALSE) {
+        auto authKeyIdBlob =
+            reinterpret_cast<CERT_AUTHORITY_KEY_ID2_INFO*>(certBuff.data());
+
+        toHexStr(authKeyIdBlob->KeyId.pbData,
+                 authKeyIdBlob->KeyId.pbData + authKeyIdBlob->KeyId.cbData,
+                 authKeyId);
+      } else {
+        VLOG(1) << "Failed to decode authority_key_id with (" << GetLastError()
+                << ")";
+      }
+    }
+  }
+  r["authority_key_id"] = authKeyId;
+
+  results.push_back(r);
+}
+
+void findUserPersonalCertsOnDisk(const std::string& username,
+                                 const std::string& storeId,
+                                 const std::string& sid,
+                                 const std::string& storeName,
+                                 const std::string& storeLocation,
+                                 QueryData& results) {
+  std::stringstream certsPath;
+  certsPath
+      << getSystemRoot().root_name().string() << "\\Users\\" << username
```

Comment:
Is this the best approach? Could this find the home directory from the `users` table? I am not sure if Windows users can change their home to another location. If they cannot then this seems fine.

Reply:
Thanks for questioning this! I refactoring things to reuse some code from the `users` table for retrieving the home directory of a user. In doing so, I discovered that system accounts (e.g. Local System) also have directories for Personal certificates, and so I've modified the PR to also be able to retrieve these. (Previously, it would bail out with a filesystem error). I've modified the PR description with a few screenshots of this.